# Pandas DataFrames
#### Dylan Manchester

Datasource: https://www.kaggle.com/lava18/google-play-store-apps
***

## Introduction

Pandas is an open source Python package that makes working with data easier.

The primary Pandas objects are Series and DataFrames.

A Series is a 1D array where the elements have index labels.

A DataFrame is a 2D array with labeled columns and indexed rows, comparable to an SQL table but more features.

We will be exploring DataFrames today, so we must first import Pandas.

We also want to import Numpy for later use.

In [1]:
import pandas as pd
import numpy as np

***

## Creating DataFrames

### DataFrame Method
Syntax: DataFrame(data=None, index=None, columns=None, dtype=None, copy=False)



#### No Data

The simplest way to create a DataFrame is to provide just provide the structure without any data.

In [2]:
df = pd.DataFrame(index=['A', 'B', 'C'])
print(df)
df

Empty DataFrame
Columns: []
Index: [A, B, C]


Empty DataFrame
Columns: []
Index: [A, B, C]

In [3]:
df = pd.DataFrame(index=['A','B','C'], columns=['W', 'X','Y','Z'])
df

W    X    Y    Z
A  NaN  NaN  NaN  NaN
B  NaN  NaN  NaN  NaN
C  NaN  NaN  NaN  NaN

#### List of Lists

When providing a list of lists, each internal list is interpreted as a row.

It is important to note that all of the lists must be the same length for this to work.

In [4]:
data=[
    ["A",1,3],
    ["B",6,8],
    ["C",3,4],
    ["D",9,12],]
df = pd.DataFrame(data)
df

0  1   2
0  A  1   3
1  B  6   8
2  C  3   4
3  D  9  12

#### Dictionary of Lists

If the lists we have represent a column rather than a row, we can use a dictionary of the lists where the key for each list becomes the column label.

This method also requires the lists to be the same length.

In [5]:
data={
    "Name":["John","Jo","Sam","April"],
    "Year":["Senior","Junior","Freshman","Junior"],
    "GPA":[3.5,3.0,4.0,3.8]
}
df = pd.DataFrame(data)
df

Name      Year  GPA
0   John    Senior  3.5
1     Jo    Junior  3.0
2    Sam  Freshman  4.0
3  April    Junior  3.8

#### List of Dictionaries

For more less structured data, a list of dictionaries may be more appropriate. 

Each dictionary represents a row and each key in the dictionary represents a column.

The dictionaries may have any number of keys and they are not required to match.

In [6]:
data=[
    {"Name":"John","GPA":3.5},
    {},
    {"Name":"Jo","Year":"Junior","Job":"Waiter","GPA":3.0},
    {"Name":"Sam","Job":"Data Analyst","GPA":4.0},
    {"Name":"April","Year":"Junior","GPA":3.8}]
df = pd.DataFrame(data)
df

Name  GPA    Year           Job
0   John  3.5     NaN           NaN
1    NaN  NaN     NaN           NaN
2     Jo  3.0  Junior        Waiter
3    Sam  4.0     NaN  Data Analyst
4  April  3.8  Junior           NaN

### Reading Data

More often than not, we will be gathering data from external sources.

We can use a wide array of different input methods to create DataFrames from specific types of sources

#### Flat Files

Flat files include text or csv files that use a delimiter to seperate the columns.

We can use the read_table and read_csv methods with the only difference being read_table assumes a tab delimiter while read_csv assumes a comma.

There is a plethora of optional parameters to explore when working with different datasets.

Check out the pandas documentation here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_table.html

In [7]:
apps = pd.read_table("googleplaystore.csv",delimiter=',')
apps

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

In [8]:
app_reviews = pd.read_csv("googleplaystore_user_reviews.csv")
app_reviews

App  \
0            10 Best Foods for You   
1            10 Best Foods for You   
2            10 Best Foods for You   
3            10 Best Foods for You   
4            10 Best Foods for You   
...                            ...   
64290  Houzz Interior Design Ideas   
64291  Houzz Interior Design Ideas   
64292  Houzz Interior Design Ideas   
64293  Houzz Interior Design Ideas   
64294  Houzz Interior Design Ideas   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
2                                                    NaN       NaN   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
...                                                  ...       ...   
64290                                                NaN       NaN   
64291                                                NaN       NaN   
64292                                                NaN       NaN   
64293                                                NaN       NaN   
64294                                                NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                    1.00                0.533333  
1                    0.25                0.288462  
2                     NaN                     NaN  
3                    0.40                0.875000  
4                    1.00                0.300000  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[64295 rows x 5 columns]

#### HTML

Another practical datasource is an HTML file.

Using a url to an html webpage containing a table, the read_html method will extract a list of tables.

The tables found within the webpage are returned as a list.

In [9]:
df = pd.read_html("https://www.nasdaq.com/market-activity/stocks/msft/historical")
df[0]

***

## Accessing Data

Now that we have covered how to get data into a DataFrame, we need to understand how to access data within a DataFrame

### Basic Indexing

This indexing method is the most convenient when looking to access a specific column or list of columns.

In [10]:
app_reviews[["App"]]

App
0            10 Best Foods for You
1            10 Best Foods for You
2            10 Best Foods for You
3            10 Best Foods for You
4            10 Best Foods for You
...                            ...
64290  Houzz Interior Design Ideas
64291  Houzz Interior Design Ideas
64292  Houzz Interior Design Ideas
64293  Houzz Interior Design Ideas
64294  Houzz Interior Design Ideas

[64295 rows x 1 columns]

In [11]:
app_reviews[["Sentiment","Sentiment_Polarity","Sentiment_Subjectivity"]]

Sentiment  Sentiment_Polarity  Sentiment_Subjectivity
0      Positive                1.00                0.533333
1      Positive                0.25                0.288462
2           NaN                 NaN                     NaN
3      Positive                0.40                0.875000
4      Positive                1.00                0.300000
...         ...                 ...                     ...
64290       NaN                 NaN                     NaN
64291       NaN                 NaN                     NaN
64292       NaN                 NaN                     NaN
64293       NaN                 NaN                     NaN
64294       NaN                 NaN                     NaN

[64295 rows x 3 columns]

### Boolean Indexing

With boolean indexing, we can return a subset of rows based on a series pf booleans of the same length.

In [12]:
bools = app_reviews["Sentiment"]=="Positive"
print(bools)
app_reviews[bools]

0         True
1         True
2        False
3         True
4         True
         ...  
64290    False
64291    False
64292    False
64293    False
64294    False
Name: Sentiment, Length: 64295, dtype: bool


App  \
0               10 Best Foods for You   
1               10 Best Foods for You   
3               10 Best Foods for You   
4               10 Best Foods for You   
5               10 Best Foods for You   
...                               ...   
64217  Housing-Real Estate & Property   
64221  Housing-Real Estate & Property   
64222  Housing-Real Estate & Property   
64223  Housing-Real Estate & Property   
64227  Housing-Real Estate & Property   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
5                                               Best way  Positive   
...                                                  ...       ...   
64217  I able set range 1cr, scroll space 0-1cr range...  Positive   
64221  Everything old stuff neither clear sold proper...  Positive   
64222  Most ads older many agents ..not much owner po...  Positive   
64223  If photos posted portal load, fit purpose. I'm...  Positive   
64227  I property business got link SMS happy perform...  Positive   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                1.000000                0.533333  
1                0.250000                0.288462  
3                0.400000                0.875000  
4                1.000000                0.300000  
5                1.000000                0.300000  
...                   ...                     ...  
64217            0.233333                0.550000  
64221            0.021591                0.259470  
64222            0.173333                0.486667  
64223            0.225000                0.447222  
64227            0.800000                1.000000  

[23998 rows x 5 columns]

We can see that 23,998 of the reviews were marked as positive.

This can also be done with only one line and no extra variable

In [13]:
app_reviews[app_reviews["Sentiment"]=="Positive"]


App  \
0               10 Best Foods for You   
1               10 Best Foods for You   
3               10 Best Foods for You   
4               10 Best Foods for You   
5               10 Best Foods for You   
...                               ...   
64217  Housing-Real Estate & Property   
64221  Housing-Real Estate & Property   
64222  Housing-Real Estate & Property   
64223  Housing-Real Estate & Property   
64227  Housing-Real Estate & Property   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
5                                               Best way  Positive   
...                                                  ...       ...   
64217  I able set range 1cr, scroll space 0-1cr range...  Positive   
64221  Everything old stuff neither clear sold proper...  Positive   
64222  Most ads older many agents ..not much owner po...  Positive   
64223  If photos posted portal load, fit purpose. I'm...  Positive   
64227  I property business got link SMS happy perform...  Positive   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                1.000000                0.533333  
1                0.250000                0.288462  
3                0.400000                0.875000  
4                1.000000                0.300000  
5                1.000000                0.300000  
...                   ...                     ...  
64217            0.233333                0.550000  
64221            0.021591                0.259470  
64222            0.173333                0.486667  
64223            0.225000                0.447222  
64227            0.800000                1.000000  

[23998 rows x 5 columns]

Boolean indexing can also be used to select a single index and it should be noted that a DataFrame is returned.

In [14]:
apps[apps.index==1057]

App Category  Rating Reviews                Size    Installs  \
1057  Rabo Banking  FINANCE     3.4   31906  Varies with device  1,000,000+   

      Type Price Content Rating   Genres   Last Updated Current Ver  \
1057  Free     0       Everyone  Finance  July 19, 2018      5.16.0   

     Android Ver  
1057  4.0 and up

### Loc Method

The loc method can be used to access a single or set of specified rows and columns.

In [15]:
apps.loc[[1057]]

App Category  Rating Reviews                Size    Installs  \
1057  Rabo Banking  FINANCE     3.4   31906  Varies with device  1,000,000+   

      Type Price Content Rating   Genres   Last Updated Current Ver  \
1057  Free     0       Everyone  Finance  July 19, 2018      5.16.0   

     Android Ver  
1057  4.0 and up

In [16]:
app_reviews.loc[:,["App"]]

App
0            10 Best Foods for You
1            10 Best Foods for You
2            10 Best Foods for You
3            10 Best Foods for You
4            10 Best Foods for You
...                            ...
64290  Houzz Interior Design Ideas
64291  Houzz Interior Design Ideas
64292  Houzz Interior Design Ideas
64293  Houzz Interior Design Ideas
64294  Houzz Interior Design Ideas

[64295 rows x 1 columns]

In [17]:
apps.loc[1057:1060]

App Category  Rating Reviews                Size  \
1057            Rabo Banking  FINANCE     3.4   31906  Varies with device   
1058  Capitec Remote Banking  FINANCE     4.3   20672  Varies with device   
1059               Itau bank  FINANCE     4.2  957973                 40M   
1060                  Nubank  FINANCE     4.7  130582                 24M   

         Installs  Type Price Content Rating   Genres    Last Updated  \
1057   1,000,000+  Free     0       Everyone  Finance   July 19, 2018   
1058   1,000,000+  Free     0       Everyone  Finance     May 8, 2018   
1059  10,000,000+  Free     0       Everyone  Finance   July 30, 2018   
1060   5,000,000+  Free     0       Everyone  Finance  August 2, 2018   

             Current Ver         Android Ver  
1057              5.16.0          4.0 and up  
1058  Varies with device  Varies with device  
1059               6.5.7          4.2 and up  
1060  Varies with device  Varies with device

Another convenient functionality of the loc method is that it adds a new index label if the one specified does not exist.

In [18]:
df = pd.DataFrame(columns=['Name','Year','GPA'])
df.loc[0]=["John","Senior",3.5]
df.loc[1]=["Jo","Junior",3.0]
df.loc['a']=["Sam","Freshman",4.0]
df.loc['b']=["April","Junior",3.8]
df.loc['x','GPA']=4.0
df.loc['x','Year']="Freshman"
df

Name      Year  GPA
0   John    Senior  3.5
1     Jo    Junior  3.0
a    Sam  Freshman  4.0
b  April    Junior  3.8
x    NaN  Freshman  4.0

### Iloc Method

If we do not want to use the index labels, we can also access the index position using the iloc method.

In [19]:
df.iloc[4,0]=["Alex"]
df

Name      Year  GPA
0   John    Senior  3.5
1     Jo    Junior  3.0
a    Sam  Freshman  4.0
b  April    Junior  3.8
x   Alex  Freshman  4.0


However, this cannot be used to add new rows to the DataFrame

In [20]:
df.iloc[5]=["James","Senior",1.0]

IndexError: single positional indexer is out-of-bounds

***

## Exploring DataFrames

### Overview


The first exploritory task is to get an overall picture of our DataFrames.


In [21]:
print("Apps Summary:")
print(apps.shape)
print(apps.dtypes)


Apps Summary:
(10841, 13)
App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object


In [22]:
print("App Reviews Summary:")
print(app_reviews.shape)
print(app_reviews.dtypes)


App Reviews Summary:
(64295, 5)
App                        object
Translated_Review          object
Sentiment                  object
Sentiment_Polarity        float64
Sentiment_Subjectivity    float64
dtype: object


We can see that there are 10,841 different apps and a total of 64,295 reviews.

All the columns in our datasets have either an object or float64 datatype.

However, some of the columns of the app data such as Reviews and Price were assigned to the object datatype when they appeared to be numerical.

Let's figure out why:

In [24]:
def floatTest(df, column):
    x = 0
    try:
        for i in df[column]:
            x=i
            float(i)
        print("Casting "+column+" to float is possible")
    except:
        print("First non-float value in the " + column +" column: "+ x)

In [25]:
floatTest(apps,"Reviews")
floatTest(apps,"Price")

First non-float value in the Reviews column: 3.0M
First non-float value in the Price column: $4.99


When we try to cast the Reviews column to a float, we can see that "3.0M" is a value in the column.

For the Price column, the first non-float value we find is "$4.99".

This is important to know and we will resolve these values before casting the columns.

Next we want to find columns with a limited number of unique values so we can save memory by casting them to the category datatype.

In [26]:
def uniqueTable(df):
    uniques = pd.DataFrame(columns=["Data Type", "Unique Values"])
    for i in df.columns:
        uniques.loc[i] = [df[i].dtype,len(df[i].unique())]
    return uniques

In [27]:
print("Number of unique values in each column of apps DataFrame:")
uniqueTable(apps)

Number of unique values in each column of apps DataFrame:


Data Type Unique Values
App               object          9660
Category          object            34
Rating           float64            41
Reviews           object          6002
Size              object           462
Installs          object            22
Type              object             4
Price             object            93
Content Rating    object             7
Genres            object           120
Last Updated      object          1378
Current Ver       object          2833
Android Ver       object            34

From this output, we can see that Category, Installs, Type, Price, Content Rating, Genres, and Android Ver are all objects with less than 200 unique values.

Aside from Price which we will cast into a float, the rest of these columns could be stored more efficiently as a category type.

In [28]:
print("Number of unique values in each column of app_reviews DataFrame:")
uniqueTable(app_reviews)

Number of unique values in each column of app_reviews DataFrame:


Data Type Unique Values
App                       object          1074
Translated_Review         object         27995
Sentiment                 object             4
Sentiment_Polarity       float64          6196
Sentiment_Subjectivity   float64          4531

The Sentiment column only has 4 unique values so it would definitely benefit from being converted into a categoty data type.

***

## Data Cleaning

The next step in the data science lifecycle is to clean our data for analysis.

### Modifying Values

First we will focus on the Price column.

We can use the replace function combined with regular expressions (regex) to remove the $ character from the price

In [29]:
apps["Price"] = apps["Price"].replace('[\$]', '', regex=True)

Retry the floatTest to see if there are other non-float values in the Price column.

In [30]:
floatTest(apps,"Price")

First non-float value in the Price column: Everyone


It looks like one of the Apps has "Everyone" as the Price.

We can use the boolean indexing to find the culprit.

In [31]:
apps[apps["Price"]=="Everyone"]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

We can see that the values are shifted over by one column probably due to this app not having a Category.

It also happens that for this row the Reviews column contains the value "3.0M" which is the same as what the floatTest found.

Let's quickly fix it by replacing shifting the values over one column and replacing the missing Category with a null value.

In [32]:
apps.loc[10472,"Rating":"Android Ver"] = apps.loc[10472,"Category":"Current Ver"].to_numpy()
apps.loc[10472,"Category"] = np.NaN
apps.loc[[10472]]

App Category Rating Reviews  Size  \
10472  Life Made WI-Fi Touchscreen Photo Frame      NaN    1.9      19  3.0M   

      Installs  Type Price Content Rating Genres       Last Updated  \
10472   1,000+  Free     0       Everyone    NaN  February 11, 2018   

      Current Ver Android Ver  
10472      1.0.19  4.0 and up

Retry the floatTest again for both columns.

In [33]:
floatTest(apps,"Price")
floatTest(apps,"Reviews")

Casting Price to float is possible
Casting Reviews to float is possible


It looks like our fixes worked!

### Changing Column Types

Now we can cast the columns using the astype method.

Using the memory_usage method before and after casting the columns to show how much memory was saved.

In [34]:
print("Memory usage of Apps Data before cast: " + str(apps.memory_usage().sum()))

Memory usage of Apps Data before cast: 1127592


In [35]:
casts = {
    "Category":"category",
    "Rating":"float64",
    "Reviews":"float64",
    "Installs":"category",
    "Type":"category",
    "Price": "float64",
    "Content Rating":"category",
    "Genres":"category",
    "Android Ver":"category"}
apps = apps.astype(casts)
apps.dtypes

App                 object
Category          category
Rating             float64
Reviews            float64
Size                object
Installs          category
Type              category
Price              float64
Content Rating    category
Genres            category
Last Updated        object
Current Ver         object
Android Ver       category
dtype: object

In [36]:
print("Memory usage of Apps Data after cast: " + str(apps.memory_usage().sum()))

Memory usage of Apps Data after cast: 682542


In [37]:
print("Memory usage of Sentiment Column before cast: " + str(app_reviews["Sentiment"].memory_usage()) + '\n')
app_reviews = app_reviews.astype({"Sentiment":'category'})
print("Memory usage of Sentiment Column after cast: " + str(app_reviews["Sentiment"].memory_usage())+'\n')
app_reviews.dtypes

Memory usage of Sentiment Column before cast: 514488

Memory usage of Sentiment Column after cast: 64527



App                         object
Translated_Review           object
Sentiment                 category
Sentiment_Polarity         float64
Sentiment_Subjectivity     float64
dtype: object

### Removing Null Values

The next cleaning process to undertake is to remove null values.

First we will count the null values in each column to see how we should address them.

In [38]:
def nulls(df):
    for col in df.columns:
        n=df[df[col].isna()]
        print(str(col)+" : "+str(len(n)))

In [39]:
nulls(apps)

App : 0
Category : 1
Rating : 1474
Reviews : 0
Size : 0
Installs : 0
Type : 1
Price : 0
Content Rating : 0
Genres : 1
Last Updated : 0
Current Ver : 8
Android Ver : 2


Some of the columns only have a few null values so we will remove the rows containing those.

However, the Rating column has numerous null values so we will create a seperate DataFrame with these values removed.

In [40]:
apps.dropna(subset=["Category","Type","Genres","Current Ver","Android Ver"],inplace=True)

In [41]:
apps_with_rating=apps.dropna()

Now we have two DataFrames: 

- apps which only has null values in Ratings 
- apps_with_rating which has no null values

In [42]:
nulls(app_reviews)

App : 0
Translated_Review : 26868
Sentiment : 26863
Sentiment_Polarity : 26863
Sentiment_Subjectivity : 26863


Many of these rows seem to have all null values except for the app name so we will remove all rows with null values.

In [43]:
app_reviews.dropna(inplace=True)

***

## Writing Data

Since our data is more usable now, we should create files to save it for future access.

Our first option is to write to a csv file with the added option of compressing it to reduce storage.

In [44]:
apps.to_csv("googleplaystoreclean.csv",mode='w',compression='zip')

Another storage format is an HDF file where we can write multiple DataFrames to the same file using a key.

In [59]:
apps_with_rating.to_hdf("appdata.hd5",key='ratings',mode='w',format='table')
app_reviews.to_hdf("appdata.hd5",key='reviews',format='table')

There are many other options for storage formats and optional parameters that can be explored here: https://pandas.pydata.org/docs/reference/frame.html#serialization-io-conversion

***

## Manipulation Functions

Another important feature of Pandas DataFrames are the manipulation functions.

### Concatinate

The concat method is for combining DataFrame either by column or by row.

Multiple DataFrames can be concatinated at a time and there are many join type and indexing options.

In [45]:
df1 = pd.DataFrame([
    [0,1,2],
    [1,2,3],
    [2,3,4]])
df2 = pd.DataFrame([
    [4,5,6],
    [6,7,8],
    [8,9,10]])
df3 = pd.DataFrame([
    [11,12,13],
    [13,14,15],
    [15,16,17]])

In [46]:
pd.concat([df1,df2,df3],axis=1,ignore_index=True)

0  1  2  3  4   5   6   7   8
0  0  1  2  4  5   6  11  12  13
1  1  2  3  6  7   8  13  14  15
2  2  3  4  8  9  10  15  16  17

In [47]:
pd.concat([df1,df2,df3],axis=0,keys=["A","B","C"])

0   1   2
A 0   0   1   2
  1   1   2   3
  2   2   3   4
B 0   4   5   6
  1   6   7   8
  2   8   9  10
C 0  11  12  13
  1  13  14  15
  2  15  16  17

### Append

Append is a function can concatinate rows, providing a subset of the concat functionality.

In [48]:
df1.append([df2,df3],ignore_index=True)

0   1   2
0   0   1   2
1   1   2   3
2   2   3   4
3   4   5   6
4   6   7   8
5   8   9  10
6  11  12  13
7  13  14  15
8  15  16  17

### Merge

The merge function can be used to combine two DataFrames based on column values, similar to the SQL join function.

The left and right parameters are the respective DataFrames while the left_on and right_on parameters dictate which columns will be used for the join.

In [49]:
data1 = {
    "Student":["John","Chris","James"],
    "Class":["A","A","C"],
    "Midterm_Grade":[50,60,80],
    "Final":[60,60,80],
}
data2 = {
    "Class":["A","B","C","D","E"],
    "Student Average":[77,88,99,87,78],
}
df1 = pd.DataFrame(data1)
df2 = pd.DataFrame(data2)
pd.merge(left=df1,right=df2,left_on="Class",right_on="Class",how="outer",validate="m:1")

Student Class  Midterm_Grade  Final  Student Average
0    John     A           50.0   60.0               77
1   Chris     A           60.0   60.0               77
2   James     C           80.0   80.0               99
3     NaN     B            NaN    NaN               88
4     NaN     D            NaN    NaN               87
5     NaN     E            NaN    NaN               78

### Join

The join method is for adding columns of one DataFrame to another based on index values which provides a subset of the functionality of Merge.

In [50]:
df1.set_index("Class").join(df2.set_index("Class"),how="outer")

Student  Midterm_Grade  Final  Student Average
Class                                               
A        John           50.0   60.0               77
A       Chris           60.0   60.0               77
B         NaN            NaN    NaN               88
C       James           80.0   80.0               99
D         NaN            NaN    NaN               87
E         NaN            NaN    NaN               78

***

## Analysis

After cleaning data, we can start our analysis.

### Most Expensive Category

This first analysis will find the 5 categories with the highest average price.

We can start reading from the compressed csv.

In [51]:
df = pd.read_csv("googleplaystoreclean.csv",index_col=[0],compression='zip',usecols=["Category","Price"])

Next we can group by Category and take the mean of each.

In [53]:
category_avg_price = df.groupby(by=["Category"],axis='index').mean()
category_avg_price

Price
Category                     
ART_AND_DESIGN       0.093281
AUTO_AND_VEHICLES    0.158471
BEAUTY               0.000000
BOOKS_AND_REFERENCE  0.520739
BUSINESS             0.402761
COMICS               0.000000
COMMUNICATION        0.214832
DATING               0.134316
EDUCATION            0.115128
ENTERTAINMENT        0.053557
EVENTS               1.718594
FAMILY               1.236682
FINANCE              7.925765
FOOD_AND_DRINK       0.066772
GAME                 0.251136
HEALTH_AND_FITNESS   0.197478
HOUSE_AND_HOME       0.000000
LIBRARIES_AND_DEMO   0.011786
LIFESTYLE            6.180288
MAPS_AND_NAVIGATION  0.196715
MEDICAL              3.110065
NEWS_AND_MAGAZINES   0.014064
PARENTING            0.159667
PERSONALIZATION      0.390949
PHOTOGRAPHY          0.400627
PRODUCTIVITY         0.591816
SHOPPING             0.021077
SOCIAL               0.054136
SPORTS               0.260417
TOOLS                0.316599
TRAVEL_AND_LOCAL     0.193605
VIDEO_PLAYERS        0.059771
WEATHER              0.395366

Sort the results by Price in descending order.

In [54]:
sorted_avg_price = category_avg_price.sort_values(by=["Price"],axis='index',ascending=False)
sorted_avg_price

Price
Category                     
FINANCE              7.925765
LIFESTYLE            6.180288
MEDICAL              3.110065
EVENTS               1.718594
FAMILY               1.236682
PRODUCTIVITY         0.591816
BOOKS_AND_REFERENCE  0.520739
BUSINESS             0.402761
PHOTOGRAPHY          0.400627
WEATHER              0.395366
PERSONALIZATION      0.390949
TOOLS                0.316599
SPORTS               0.260417
GAME                 0.251136
COMMUNICATION        0.214832
HEALTH_AND_FITNESS   0.197478
MAPS_AND_NAVIGATION  0.196715
TRAVEL_AND_LOCAL     0.193605
PARENTING            0.159667
AUTO_AND_VEHICLES    0.158471
DATING               0.134316
EDUCATION            0.115128
ART_AND_DESIGN       0.093281
FOOD_AND_DRINK       0.066772
VIDEO_PLAYERS        0.059771
SOCIAL               0.054136
ENTERTAINMENT        0.053557
SHOPPING             0.021077
NEWS_AND_MAGAZINES   0.014064
LIBRARIES_AND_DEMO   0.011786
COMICS               0.000000
BEAUTY               0.000000
HOUSE_AND_HOME       0.000000

And output the top 5.

In [55]:
analysis1 = sorted_avg_price.iloc[0:5]
analysis1

Price
Category           
FINANCE    7.925765
LIFESTYLE  6.180288
MEDICAL    3.110065
EVENTS     1.718594
FAMILY     1.236682

Since each step uses a method of the datatype returned by the prior step, the entire analysis can be done in one line.

In [56]:
pd.read_csv("googleplaystoreclean.csv",index_col=[0],usecols=["Category","Price"],compression='zip').groupby(by=["Category"],axis='index').mean().sort_values(by=["Price"],axis='index',ascending=False).iloc[0:5]

Price
Category           
FINANCE    7.925765
LIFESTYLE  6.180288
MEDICAL    3.110065
EVENTS     1.718594
FAMILY     1.236682

The analysis results can be written to an html file.

In [57]:
analysis1.to_html("Top5MostExpensiveCategories.html")

### Well Liked Apps

The second analysis will determine all apps that have a rating of 4.5 or higher and rank them by Sentiment.

We can start by reading from the HDF file.

In [60]:
rating = pd.read_hdf("appdata.hd5",key="ratings",columns=["App","Rating"])
review = pd.read_hdf("appdata.hd5",key="reviews",columns=["App","Sentiment"])

Next we can use boolean indexing to select only the apps with ratings of at least 4.5.

In [61]:
rating = rating[rating["Rating"]>=4.5]

We can replace the Sentiment category with numerical values.

In [62]:
review["Sentiment"].dtypes

CategoricalDtype(categories=['Negative', 'Neutral', 'Positive'], ordered=False)

In [63]:
review.replace({"Positive":1,"Neutral":0,"Negative":-1},inplace=True)

Then we can merge the DataFrames to have all the necessary information in one DataFrame.

In [64]:
rating_review = rating.merge(right=review,how="inner",left_on="App",right_on="App")
rating_review

App  Rating  Sentiment
0      Colorfit - Drawing & Coloring     4.7          1
1      Colorfit - Drawing & Coloring     4.7         -1
2      Colorfit - Drawing & Coloring     4.7          1
3      Colorfit - Drawing & Coloring     4.7         -1
4      Colorfit - Drawing & Coloring     4.7          1
...                              ...     ...        ...
26878   A+ Gallery - Photos & Videos     4.5          1
26879   A+ Gallery - Photos & Videos     4.5          1
26880   A+ Gallery - Photos & Videos     4.5          1
26881   A+ Gallery - Photos & Videos     4.5          0
26882   A+ Gallery - Photos & Videos     4.5          1

[26883 rows x 3 columns]

We can now find the mean of the Sentiment grouped by App.

The results will be sorted by average Sentiment score and then by Rating.

In [65]:
analysis2 = rating_review.groupby(by="App",sort=False).mean().sort_values(by=["Sentiment","Rating"],ascending=False)
analysis2

Rating  Sentiment
App                                             
Down Dog: Great Yoga Anywhere     4.9   1.000000
GPS Speedometer and Odometer      4.8   1.000000
Brightest Flashlight Free ®       4.7   1.000000
Calculator - unit converter       4.7   1.000000
Daniel Tiger for Parents          4.7   1.000000
...                               ...        ...
Cooking Fever                     4.5  -0.177778
Call Blocker                      4.6  -0.333333
Free Live Talk-Video Call         4.7  -1.000000
Discover Mobile                   4.6  -1.000000
Fruit Block - Puzzle Legend       4.6  -1.000000

[263 rows x 2 columns]

Another way we could find the mean of the Sentiment grouped by App is to use the pivot_table method.

Using App as the index, we are selecting that as the group by variable and the default aggregation function is mean so we do not have to specify.

In [66]:
analysis2_pivot = rating_review.pivot_table(index="App").sort_values(by=["Sentiment","Rating"],ascending=False)
analysis2_pivot.eq(analysis2).all()

Rating       True
Sentiment    True
dtype: bool

We can write the completed analysis back to the HDF file.

In [67]:
analysis2.to_hdf("appdata.hd5",key='analysis')

***